merkle.c

```diff
@@ -0,0 +1,308 @@
+                *intermediate_nodes[i - 1] = parent = merk_alloc_node();
+
+                *parent = (merkle_node_t) {
+                    .ptr = sibling->ptr,
+                    .left = node,
+                    .right = sibling,
+                };
+            } else if (node->ptr > sibling->ptr) {
+                *intermediate_nodes[i - 1] = parent = merk_alloc_node();
+
+                *parent = (merkle_node_t) {
+                    .ptr = node->ptr,
+                    .left = sibling,
+                    .right = node,
+                };
+            } else {
+                // We've specified a key that already exists, so overwrite the old node.
+                i--;
+                *intermediate_nodes[i] = node;
+                merk_free_node(sibling);
+            }
+            break;
+        }
+
+        // Traverse the BST
+
+        if (node->ptr < parent->ptr) {
+            if (!parent->left) {
+                parent->left = node;
+                break;
+            } else {
+                intermediate_nodes[i] = &parent->left;
+            }
+        } else {
+            if (!parent->right) {
+                parent->right = node;
+                break;
+            } else {
+                intermediate_nodes[i] = &parent->right;
+            }
+        }
+    }
+
+    if (i == MERK_MAX_DEPTH) {
+        printf("Inserted to merkle tree with problematic key insertion order! "
+               "This has led to an unbalanced tree exceeding the depth capacity of %d. "
+               "Aborting!", MERK_MAX_DEPTH);
+        assert(false);
+    }
+
+    for (i = i - 1; i >= 0; i--) {
+        // Here we walk back up the tree to percolate up our new hashes.
+        // We keep the previous iteration's hash, as well as the current iteration's merkle node,
+        // in secure memory to avoid any race conditions after writing to DRAM in the last step.
+        //
+        // We otherwise aren't concerned that an attacker will modify any data
+        // in the tree during this stage, as doing so will compromise the integrity
+        // of the tree such that the user will be alerted upon attempting any accesses
+        // to the compromised location.
+        //
+        // We also mark accesses to parent_ptr as volatile to ensure they get written and read
+        // with the correct access pattern.
+
+        sha256_init(&hasher);
+        merkle_node_t *parent_ptr = *intermediate_nodes[i];
+        merkle_node_t parent = *(volatile merkle_node_t *)parent_ptr;
+        assert(!memcmp(lowest_hash, node->hash, 32)); // TODO: this sanity checking can probably go away
+
+        // Check which child we are, and compute the parent's hash with our pre-stored `lowest_hash`
+        // and the hash of our sibling, if it exists.
+        if (node == parent.left) {
+            sha256_update(&hasher, lowest_hash, 32);
+            if (parent.right)
+                sha256_update(&hasher, parent.right->hash, 32);
```

Comment:
I've always been a little suspicious of myself doing this but somehow managed to justify it away in my head. But it's wrong. Reaching over to grab the sibling's hash when we're updating the tree is implicitly trusting that the sibling hasn't been tampered with by an adversary. I think that as I walk down the tree, I need to also verify the entries and save the sibling hashes. Unfortunately, that increases our space requirements to 40 bytes per layer of tree.